actions/websocket_actions.jsx

```diff
@@ -1448,6 +1451,25 @@ function handleRefreshAppsBindings() {
     };
 }
 
+export function handleAppsPluginEnabled() {
+    return (doDispatch, doGetState) => {
+        doDispatch({
+            type: AppsTypes.APPS_PLUGIN_ENABLED,
+        });
+
+        const state = doGetState();
+        doDispatch(fetchAppBindings(getCurrentUserId(state), getCurrentChannelId(state)));
```

Comment:
Probably not a big deal, but can we batch these dispatches?

Reply:
@larkox I decided not to batch the actions because:

- it made testing that bindings were being fetched a little more complicated
- having separate actions dispatched makes it easier to debug in redux dev tools

Reply:
Now that I think about this deeper, we shouldn't do this. This would imply that the moment we enable the plugin, all active users will ask the server for the plugin bindings, and that is costly. Since enabling or disabling the plugin is not something "critical" for the end user, we can wait until the user switch channels to get the bindings. What do you think?

Reply:
@larkox Thinking about it more with your input, we can simply mark the plugin status as enabled in the reducer when the plugin is enabled, and let the user carry on without immediately fetching new bindings.

This is a similar situation to how we operate with webapp plugins. When a given plugin is enabled, every user fetches that plugin's webapp bundle simultaneously. A big difference here is that this operation is just fetching javascript/css files. The bindings call has much more dynamic computations involved.

- When an App is installed, should we not tell all users to refresh bindings?
- Should ee should allow the frontend to request bindings for a specific App so as to not storm all of the Apps at anytime we have an "all users" refresh situation
- 2/5 Maybe we should _always_ avoid making any situation where all users refresh bindings at the same time. So no App has to be able to handle all of those requests at the same time.

Something else that comes to mind: If an App has no channel-specific bindings, we can tell the Apps plugin to not fetch that App's bindings on channel switches. The frontend can create "soft" bindings requests for channel switches, and the Apps plugin would only fetch bindings from the Apps that are registered to have channel-specific bindings. The frontend also creates "hard" bindings requests, when it knows that it needs to do a full refresh of bindings. Then the Apps plugin would fetch from all Apps.

To accomplish this, we would need to include information in the call response about all Apps that are enabled, and which bindings can be cached that the frontend already has access to. The frontend also needs to know when to invalidate these bindings. We can send a hash of the cached response in the bindings request to the plugin, and have the plugin check if that hash is the last response computed for that user. The App or Apps plugin would be responsible for creating and storing this hash. The Apps plugin will read the stored hash and compare with what the frontend is providing in its request. What do you think?

cc @hanzei

Reply:
I agree on not asking for bindings from all users at the same time.

Regarding caching, we've talked a lot about that in the past, and everything has its ups and downs. Mainly, we need to trust the app developer to be conscious about what of their bindings are cacheable, and give them a way to invalidate the caches, which is putting a lot more of effort on the app developer side for an implementation detail (cache).

Reply:
@levb Do you have any thoughts on the caching strategy discussed in the last two paragraphs of my comment above?

Reply:
Leaving the caching discussion aside, I'm wondering how much of an issue it is, when all clients as fetch bindings "at the same time".

First of call, I would like to point out that "at thee same time" can span over multiple seconds. Clients needs to receive the websocket event (variant delay), process and respond (variant delay) to it. Depending on the network this can take a couple of seconds.
For requests to the mm server and the apps plugins, the load doesn't seem like an issue. It's the same as for enabling a plugin. For the app itself processing all of these requests might be an issue, but the are naturally distributed for a few seconds. At least serverless infrastructure like AWS can handle this. HTTP apps might be an issue, but it's a deployment type that requires a load balancer anyway if run in production.

Reply:
@hanzei Can we discuss this in our meeting this morning?